### DISCUSSION TITLE
Need help using findRecordsByExpr / findRecordsByFilter

### DISCUSSION THREAD

**Discussion by Ilyas-Codes** [pocketbase/pocketbase#4305] [Q&A] [answered]
Hello everyone, so in my `onRecordAfterUpdateRequest` hook, I'd like to find all the records in my `progress` collection that have their `assignee` field matching the removed assignees, and course field matching the `updatedRecord.id` and then delete those records. I tried both `findRecordsByExpr` and `findRecordsByFilter` but couldn't get it to work.

Here is the hook:
```
onRecordAfterUpdateRequest((e) => {
  const updatedRecord = e.record;
  const originalRecord = updatedRecord.originalCopy();
  const originalAssignees = originalRecord.getStringSlice("assignees");
  let updatedAssignees = updatedRecord.getStringSlice("assignees");
  const progressCollection = $app.dao().findCollectionByNameOrId("progress");
  const usersCollection = $app.dao().findCollectionByNameOrId("users");
  const assign_to_everyone = updatedRecord.getBool("assign_to_everyone");

  // if assign_to_everyone is true, add all the users to the assignees field
  if (assign_to_everyone) {
    const allUsers = $app.dao().findRecordsByExpr(usersCollection.name);
    updatedAssignees = allUsers.map((user) => user.id);
    updatedRecord.set("assignees", updatedAssignees);
    $app.dao().saveRecord(updatedRecord);
  }

  // users added to the assignees field
  const newAssignees = updatedAssignees.filter(
    (assignee) => !originalAssignees.includes(assignee)
  );

  // users removed from the assignees field
  const removedAssignees = originalAssignees.filter(
    (assignee) => !updatedAssignees.includes(assignee)
  );

  // create progress records for every assignee added
  newAssignees.forEach((assignee) => {
    const progressRecord = new Record(progressCollection, {
      course: updatedRecord.id,
      assignee: assignee,
      status: "Not Started",
    });

    $app.dao().saveRecord(progressRecord);
  });

  // delete progress records for every assignee removed
  removedAssignees.forEach((assignee) => {
    // this is where I probably messed up
    const progressRecordsToDelete = $app
      .dao()
      .findRecordsByExpr(
        progressCollection.name,
        $dbx.exp({ assignee: assignee, course: updatedRecord.id })
      );

    progressRecordsToDelete.forEach((progressRecord) => {
      $app.dao().deleteRecord(progressRecord);
    });
  });
}, "courses");
```

**Comment by ganigeorgiev** [comment]
> but couldn't get it to work

What actually is the issue? Is there any error returned?

That's a lot of code without much details about your collections structure or desired outcome.

**Reply by ganigeorgiev** [reply]
Also the arguments for `$dbx.exp()` are incorrect as the first one must be the raw expression. If you want to specify a key-value matching expression you need to use `$dbx.hashExp`.

For more details you could refer to https://pocketbase.io/docs/js-database/#where-andwhere-orwhere.

**Reply by Ilyas-Codes** [reply, marked as answer]
Hey Gani, thank you for the quick reply
never mind actually, I figured it out I, was using the wrong syntax for the filter expression, this solved it for me:

```
  // delete progress records for every assignee removed
  removedAssignees.forEach((assignee) => {
    const filterExpression = `assignee = '${assignee}' && course = '${updatedRecord.id}'`;
    const progressRecordsToDelete = $app
      .dao()
      .findRecordsByFilter(progressCollection.name, filterExpression);

    progressRecordsToDelete.forEach((progressRecord) => {
      $app.dao().deleteRecord(progressRecord);
    });
  });
```

**Reply by ganigeorgiev** [reply]
_As a side-note, if you need a simple cascade delete behavior you can enable it from the relation field options in the Admin UI._

**Reply by Ilyas-Codes** [reply]
ah yes, I know about that feature, really helps. however in this case I need to delete a record from my `progress` collection whenever a user is removed from an `assignees` field in my courses collection, and not when the entire record is deleted. sorry I know I'm not explaining this very well lol. but anyway cascade delete won't work here since it only deletes records from a collection when a corresponding record from a different collection is deleted but not when a field is updated.
